File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        
      int n=nums.size();
        vector<vector<int>> dp(n+1,vector<int> (n+1,0));
        
        for(int i=n-1;i>=0;i--){
            for(int prev_ind=i-1;prev_ind>=-1;prev_ind--){
                int taking=0;
                int nottaking=0;
                if(prev_ind==-1 || nums[i]>nums[prev_ind]){
                    taking=dp[i+1][i+1]+1;
                }
                nottaking=dp[i+1][prev_ind+1];

                dp[i][prev_ind+1]=max(taking,nottaking);
            }
        }
        
        return dp[0][0];
        
    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
               int[] dp = new int[nums.length];
        //considering max=1 as each index value in itself is a sunsequence of array
        int max = 1;

        //outer loop for actual array
        for(int i=0;i<nums.length;i++){
            //initializing base LIS at index
            dp[i] = 1;
            //checking from 0 to <i 
            for(int j=0;j<i;j++){
                //if number is < currValue of nums[i]
                if(nums[j]<nums[i]){
                    dp[i] = Math.max(dp[i],dp[j]+1);
                    max = Math.max(max,dp[i]);
                }
            }
        }
        return max;
    }
}